modules/metastorage-client/src/main/java/org/apache/ignite/internal/metastorage/client/MetaStorageServiceImpl.java

```diff
@@ -0,0 +1,366 @@
+import org.apache.ignite.internal.metastorage.common.command.GetAndRemoveAllCommand;
+import org.apache.ignite.internal.metastorage.common.command.GetAndRemoveCommand;
+import org.apache.ignite.internal.metastorage.common.command.GetCommand;
+import org.apache.ignite.internal.metastorage.common.command.PutAllCommand;
+import org.apache.ignite.internal.metastorage.common.command.PutCommand;
+import org.apache.ignite.internal.metastorage.common.command.RangeCommand;
+import org.apache.ignite.internal.metastorage.common.command.RemoveAllCommand;
+import org.apache.ignite.internal.metastorage.common.command.RemoveCommand;
+import org.apache.ignite.internal.metastorage.common.command.WatchExactKeysCommand;
+import org.apache.ignite.internal.metastorage.common.command.WatchRangeKeysCommand;
+import org.apache.ignite.lang.IgniteLogger;
+import org.apache.ignite.lang.IgniteUuid;
+import org.apache.ignite.metastorage.client.MetaStorageService;
+import org.apache.ignite.metastorage.common.Condition;
+import org.apache.ignite.metastorage.common.Cursor;
+import org.apache.ignite.metastorage.common.Entry;
+import org.apache.ignite.metastorage.common.Key;
+import org.apache.ignite.metastorage.common.Operation;
+import org.apache.ignite.metastorage.common.WatchEvent;
+import org.apache.ignite.metastorage.common.WatchListener;
+import org.apache.ignite.raft.client.service.RaftGroupService;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * {@link MetaStorageService} implementation.
+ */
+public class MetaStorageServiceImpl implements MetaStorageService {
+    /** The logger. */
+    private static final IgniteLogger LOG = IgniteLogger.forClass(MetaStorageServiceImpl.class);
+
+    /** Meta storage raft group service. */
+    private final RaftGroupService metaStorageRaftGrpSvc;
+
+    // TODO: IGNITE-14691 Temporally solution that should be removed after implementing reactive watches.
+    /** Watch processor, that uses pulling logic in order to retrieve watch notifications from server. */
+    private final WatchProcessor watchProcessor;
+
+    /**
+     * @param metaStorageRaftGrpSvc Meta storage raft group service.
+     */
+    public MetaStorageServiceImpl(RaftGroupService metaStorageRaftGrpSvc) {
+        this.metaStorageRaftGrpSvc = metaStorageRaftGrpSvc;
+        this.watchProcessor = new WatchProcessor();
+    }
+
+    /** {@inheritDoc} */
+    @Override public @NotNull CompletableFuture<Entry> get(@NotNull Key key) {
+        return metaStorageRaftGrpSvc.run(new GetCommand(key));
+    }
+
+    /** {@inheritDoc} */
+    @Override public @NotNull CompletableFuture<Entry> get(@NotNull Key key, long revUpperBound) {
+        return metaStorageRaftGrpSvc.run(new GetCommand(key, revUpperBound));
+    }
+
+    /** {@inheritDoc} */
+    @Override public @NotNull CompletableFuture<Map<Key, Entry>> getAll(Collection<Key> keys) {
+        List<Key> keysWithPreservedOrder = new ArrayList<>(keys);
```

Comment:
Why wee need a preserved order here and below? The order doesn't matter. We can just get `Entry.key()` from result.